syntax = "proto3";
package shiva.core.communication_objects;

message ListOfInts {
    repeated int32 data = 1;
}

message ListOfFloats {
    repeated float data = 1;
}

enum EnvironmentCommand {
    STEP = 0;
    RESET = 1;
    CLOSE = 2;
}

message EnvStepInput {
    message ListAgentsActions {
        message SingleData {
            oneof data {
                ListOfInts values_ints = 1;
                ListOfFloats values_floats = 2;
            }
        }
        repeated SingleData data = 1; // this is made to enforce either discrete actions of ints or continuous floats, not both!
    }
    map<string, ListAgentsActions> agent_actions = 1;
    EnvironmentCommand command = 2;
}

//

message EnvironmentSpecs {
    int32 observation_space = 1;
    int32 action_space = 2;
    int32 num_instances = 3;
    int32 num_agents_per_instance = 4;
}

message AgentMetrics {
    float steps_per_episode = 1;
    float step_count = 2;
    float temp_done_counter = 3;
    float done_count = 4;
    float reward_per_step = 5;
    float reward_per_episode = 6;
    float reward_total = 7;
}

message AgentState {
    int32 id = 1;
    ListOfFloats next_observation = 2;
    float reward = 3;
    bool done = 4;
}

message EnvStepOutput {
    message ListAgentsStates {
        repeated AgentState data = 1;
    }
    message ListAgentsMetrics {
        repeated AgentMetrics data = 1;
    }
    map<string, ListAgentsStates> agent_states = 1;
    map<string, ListAgentsMetrics> agent_metrics = 2;
}


service Environment {
    rpc Step (EnvStepInput) returns (EnvStepOutput) {}
}